### PR TITLE
3748 quipu actions

### PR DESCRIPTION
**Pull request by alysonturing**
### Actions:
- create-contact
- create-invoice
- create-ticket

### CODE REVIEW

File: components/quipu/actions/create-contact/create-contact.mjs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  props: {
+    app,
+    name: {
+      type: "string",
+      label: "Name",
+      description: "Contact full name.",
+    },
+    taxId: {
+      type: "string",
+      label: "Tax Id (CIF/NIF)",
```
Comment (lcaresia) [inline components/quipu/actions/create-contact/create-contact.mjs:18]:
```suggestion
      label: "Tax ID (CIF/NIF)",
```

File: components/quipu/actions/create-contact/create-contact.mjs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      "zip_code": this.zipCode,
+      "country_code": this.countryCode,
+      "client_number": this.clientNumber,
+      "supplier_number": this.supplierNumber,
+      "is_supplier_of_direct_goods": this.isSupplierOfDirectGoods,
+      "bank_account_number": this.bankAccountNumber,
+      "bank_account_swift_bic": this.bankAccountSwift,
+    };
+    const contact = await this.app.createContact($, attributes);
+    $.export("$summary", `Successfully created contact "${contact.id}"`);
```
Comment (lcaresia) [inline components/quipu/actions/create-contact/create-contact.mjs:103]:
```suggestion
    $.export("$summary", `Successfully created contact with ID "${contact.id}"`);
```

File: components/quipu/actions/create-contact/create-contact.mjs [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import app from "../../quipu.app.mjs";
+
+export default {
+  key: "quipu-create-contact",
+  name: "Creating a contact",
```
Comment (lcaresia) [inline components/quipu/actions/create-contact/create-contact.mjs:5]:
```suggestion
  name: "Create Contact",
```

File: components/quipu/actions/create-invoice/create-invoice.mjs [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import app from "../../quipu.app.mjs";
+import { ConfigurationError } from "@pipedream/platform";
+
+export default {
+  key: "quipu-create-invoice",
+  name: "Creating an invoice",
```
Comment (lcaresia) [inline components/quipu/actions/create-invoice/create-invoice.mjs:6]:
```suggestion
  name: "Create Invoice",
```

File: components/quipu/actions/create-invoice/create-invoice.mjs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        },
+      };
+    }
+    if (this.items) {
+      let items = this.items;
+      if (typeof (this.items) === "string") {
+        items = JSON.parse(items);
+      }
+      if (!Array.isArray(items)) {
+        console.log(items);
```
Comment (lcaresia) [inline components/quipu/actions/create-invoice/create-invoice.mjs:103]:
```suggestion
```

File: components/quipu/actions/create-ticket/create-ticket.mjs [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import app from "../../quipu.app.mjs";
+import { ConfigurationError } from "@pipedream/platform";
+
+export default {
+  key: "quipu-create-ticket",
+  name: "Creating a ticket",
```
Comment (lcaresia) [inline components/quipu/actions/create-ticket/create-ticket.mjs:6]:
```suggestion
  name: "Create Ticket",
```

File: components/quipu/actions/create-ticket/create-ticket.mjs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      if (!Array.isArray(items)) {
+        console.log(items);
+        throw new ConfigurationError("Items must to be an array.");
+      }
+      relationships.items = {
+        "data": items,
+      };
+    }
+    const ticket = await this.app.createTicket($, attributes, relationships);
+    $.export("$summary", `Successfully created ticket "${ticket.id}"`);
```
Comment (lcaresia) [inline components/quipu/actions/create-ticket/create-ticket.mjs:96]:
```suggestion
    $.export("$summary", `Successfully created ticket with ID "${ticket.id}"`);
```

File: components/quipu/actions/create-invoice/create-invoice.mjs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        throw new ConfigurationError("Items must to be an array.");
+      }
+      relationships.items = {
+        "data": items,
+      };
+    }
+    const invoice = await this.app.createInvoice($, attributes, relationships);
+    $.export("$summary", `Successfully created invoice "${invoice.id}"`);
```
Comment (lcaresia) [inline components/quipu/actions/create-invoice/create-invoice.mjs:111]:
```suggestion
    $.export("$summary", `Successfully created invoice with ID "${invoice.id}"`);
```

File: components/quipu/package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    "quipu"
+  ],
+  "homepage": "https://pipedream.com/apps/quipu",
+  "author": "Pipedream <support@pipedream.com> (https://pipedream.com/)",
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@pipedream/platform": "^1.2.0"
```
Comment (lcaresia) [inline components/quipu/package.json:17]: https://www.npmjs.com/package/@pipedream/platform

The lastest version is `1.2.1`, you can update to last one

File: components/quipu/quipu.app.mjs [hunk 10]
```diff
@@ -1,11 +1,146 @@
...
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    _getBaseUrl() {
+      return "https://getquipu.com";
+    },
+    _getHeaders() {
+      return {
+        "accept": "application/vnd.quipu.v1+json",
+        "content-type": "application/vnd.quipu.v1+json",
+        "Authorization": `Bearer ${this.$auth.oauth_access_token}`,
```
Comment (lcaresia) [inline components/quipu/quipu.app.mjs:64]: Move this to a getter method.

```suggestion
        "Authorization": `Bearer ${this._accessToken()}`,
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Need just some fixes!

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by DilanAthukorala** [comment]
@alysonturing following issues observed,

1. Can we add how to obtain the APP ID and Secret here?
image


2. Create Ticket Action, Items are not getting created in quipu 
image


3. Rename Components create ticket and create invoice to - "Create Income Ticket" & "Create Income Invoice" (This can help separate the expense ticket and invoices)

**Comment by alysonturing** [comment]
The first issue must to be addressed to @dannyroosevelt 

I'm fixing the rest, thanks

**Comment by dannyroosevelt** [comment]
>Can we add how to obtain the APP ID and Secret here?

Done, thanks for flagging @DilanAthukorala

**Comment by alysonturing** [comment]
@DilanAthukorala everything was resolved

**Comment by DilanAthukorala** [comment]
Verification completed, ready for release.

**Review by lcaresia — APPROVED** [review]
LGTM!

**Pull request merged**